### DISCUSSION TITLE
Porting from python 3.9 to micropython

### DISCUSSION THREAD

**Discussion by AutoPilotjn** [micropython/micropython#9327] [General]
Dear all,
First of all congratulations for the huge work done, it is impressive.
I have to port a medium size application (#10.000 lines of python code) written
in python 3.9 and up.
When porting the 3.9-3.10 code to micropython I am facing these main issues:
 - most of the data structures are decorated with @dataclass
 - use of Generic in type hints
 - use of interfaces through ABC class inheritance and abstractmethod.
 - there are some Literal declarations, but I can manage to get rid of them

The target is an STM32F7xx with 2Mbytes of Flash and 512Kbytes of Ram. The code is
validated on a Host.
I would prefer not downgrading the code: replacing dataclass with dictionnaries
and removing Generics. In addition, interfaces through ABC and abstractmethod
is a must (btw I added class ABC(): pass in abc.py). mypy is a mandatory step in our
validation process.
If it makes sense, with pointers and guidelines I could implement a subset of dataclass
including few options such as slots=True. Among difficulties, I am aware there are
cascading dependancies such as the re-regex module.
What would you advise ?

**Comment by robert-hh** [comment]
Unless most of the code lines in your package are comments or docstrings, the RAM & flash size of the target may be too small to handle a 10.000 LOC script, not considering the data size and the language incompatibilities. Pre-compiling code may help somewhat.

**Reply by dpgeorge** [reply]
10k lines of Python code should be fine if it's frozen.  Even as precompiled `.mpy` it should be OK.  (I've seen larger projects.)

**Comment by dpgeorge** [comment]
I can imaging there could be compatibility issues with new syntax features that are not implemented in MicroPython.  Also limitations of the class model (eg multiple inheritance, inheriting from a native type).

Can you give explicit examples of code that does not compile/run under MicroPython, for all 4 of your points?  I would have thought that generic type hints would work.

**Reply by AutoPilotjn** [reply]
Hereafter explicit examples extracted from code:
- dataclass
```
@dataclass(slots=True, frozen=True)
class PST_STRUCTURE:
    """
    Structure definition for Period and Start Time (PST)

    Each thread has its own definition for Real Time control
    """
    pst_name: str
    days_period: int
    hours_period: int
    minutes_period: int
    seconds_period: int | float
    hour_start_time: int
    minute_start_time: int
    second_start_time: int
```
- Generic
```
T = TypeVar("T")  # T can be 'str' or 'int'

@dataclass(slots=True)
class GET_VALUE(Generic[T]):
    value: T
    value_available: bool = False
```

```
T = TypeVar("T")  # T can be 'str' or 'int'

class FILE_ATTENTE(LOCK_, Generic[T]):
    def __init__(self, default_value: T, ...
    ...
        self.get_value: GET_VALUE[T] = GET_VALUE[T](
            value=default_value,
            value_available=False,
        )
```
- ABC classes
```
from abc import ABC, abstractmethod
class Stamp_dateable(ABC):
    `@abstractmethod`
    def stamp_date(self) -> str:
        pass

class Stamp_date_xnix(Stamp_dateable):
    """Linux and Macos implementation"""
    def stamp_date(self) -> str:
        do this ...

class Stamp_date_upy(Stamp_dateable):
    """uPython implementation"""
    def stamp_date(self) -> str:
        do that ...
```
 - Literal
```
t_OK_KO = Literal[
    "Ok",
    "Ko",
]

def ap_command_status(self) -> t_OK_KO:
        result: t_OK_KO = "Ko"
        if ...
```

**Reply by robert-hh** [reply]
Code has to be enclosed in three backticks for proper formatting. The workflow I use is.
- Past the code
- Select it
- Push Ctrl-E or click the code tag button

Then it will be enclosed in three backticks.

**Reply by AutoPilotjn** [reply]
Thanks, updated

**Reply by dpgeorge** [reply]
All of these examples will compile under MicroPython, ie they will not raise a SyntaxError.

So the issue is implementing the functionality of the classes used, eg `dataclass`, `TypeVar`, `ABC`, etc.  The should be able to be implemented in pure Python, although it may be quite tricky to get them all working correctly.

The `Literal`, `Generic` and `TypeVar` shouldn't be too hard.  For the others you can give it a try, making versions that work well enough for your application.  Otherwise you'll need to not use these features.

**Reply by AutoPilotjn** [reply]
The purpose of our project is to static type check with mypy on the host (macOS). Then to execute the same code (including the static annotations) after “compiling” with micropython on the target.
I tested the following:

```
G_Type = TypeVar("G_Type")

class GET_VALUE(Generic[G_Type]):
    pass
```

TypeVar is implemented as:
```
class TypeVar:
    pass
```

Generic is implemented as:
```
class Generic:
    pass
```

The compiling result shows that TypeVar does not raise any errors. But Generic seems to be an issue :
> TypeError: 'type' object isn't subscriptable

 Also:
```
t_OK_KO = Literal["Ok", "Ko"]
```
Literal is implemented as:
```
def Literal(*parameters):
    pass
```
And raises:
> TypeError: 'function' object isn't subscriptable

In addition declarations such as:
```
TRUE: Final[str] = "1"
```
compiles without error with micropython.
Therefore, I am in doubt whether the ‘[]’ syntax - instead of the ‘()’ syntax - would disturb micropython when used NOT as type hints but as expressions ?

**Reply by dpgeorge** [reply]
You can implement the `[]` syntax on a class, for example:
```py
class GenericClass:
    def __getitem__(self, key):
        pass

Generic = GenericClass()
```

**Reply by AutoPilotjn** [reply]
I have implemented the suggested '__getitem__()' :
Nota: I have also been inspired by https://github.com/pfalcon/pycopy-lib/blob/master/typing/typing.py .
```
class _Subscriptable:
    def __getitem__(self, sub):
        pass

_SubSingleton = _Subscriptable()

Literal = _SubSingleton
Generic = _SubSingleton
```
Literal statements does not throws error, neither TypeVar
```
t_OK_KO = Literal["Ok", "Ko"]

def TypeVar(name, *constraints, bound=None, covariant=False, contravariant=False):
    return "TypeVar string"  # or return None, same issue with Generic, see below

G_Type = TypeVar("G_Type")
```
However the following code throws an error at run-time
```
class GetValue(Generic[G_Type]):
    pass
```
```
TypeError: can't create 'NoneType' instances
```
When declaring a '__class_getitem__()':
```
class _SubscriptableClass:
    def __class_getitem__(self, sub):
        return None
_SubSingletonClass = _SubscriptableClass()
Generic = _SubSingletonClass
```
I get the following run-time error:
```
TypeError: '_SubscriptableClass' object isn't subscriptable
```
In both cases '__getitem__()' and '__class_getitem__()' when removing the generic type parameter:
```
class GetValue(Generic):
    pass
```
There is no error.

I am a bit lost. I read in 
> https://github.com/micropython/micropython-lib/issues/190 (dated [8 Aug 2018])
> [...] This is because Micropython does not support the __class_getitem__(cls, key) special function that allows the creation of generic class objects. @dpgeorge would it be possible to add that feature?

Any pointers would be very much appreciated.

**Reply by cwalther** [reply]
Your `_Subscriptable` is equivalent to
```py
class _Subscriptable:
    def __getitem__(self, sub):
        return None
```
and that is the `None` that it’s complaining about (you are effectively trying `class GetValue(None):` which doesn’t work becaue `None` is not a valid base class).

Try this instead:
```py
class _Subscriptable:
    def __getitem__(self, sub):
        return object
```

**Reply by AutoPilotjn** [reply]
Thanks, works with 'object'.
I have now an updated dummy implementation of 'typing.py'. Shall I share it ?
Please note I am a newbie to git and github, but willing to learn.

**Reply by dpgeorge** [reply]
> I have now an updated dummy implementation of 'typing.py'. Shall I share it ?

If you'd like to open a PR at https://github.com/micropython/micropython-lib with it, we can consider merging it there.

**Comment by AutoPilotjn** [comment]
Update.
It was not possible get rid of advanced type hints such as Generic[T] in a fake typing module, as class_getitem() is not implemented in MicroPython (no metaclass mecanism implemented).
Therefore the solution was to clean all type hints in the source code by the means of AST module, as described in https://stackoverflow.com/questions/42733877/remove-type-hints-in-python-source-programmatically.
Then compiling the stripped code with MicroPython works fine.
The development workflow is as follows:
1.	The typed source code is written with VS code and Pylance running in the background for checking the type hints across all code base (as well as Mypy)
2.	The typed source code is tested on a host with CPython V3.9 or up and dedicated modules simulating IOs. During this step, it shall be taken care of MicroPython differences from CPython
3.	The stripped code (without type hints) runs on the host with the Unix MicroPython compiler with dedicated modules simulating IOs
4.	The stripped code is then loaded and run on the STMF7xx target

This brings all innovative features of the Python ecosystem, before running on the target.

**Reply by winni2k** [reply]
This sounds great! I would also like to see dataclasses as part of micropython. Any chance you could share your implementation?

**Reply by AutoPilotjn** [reply]
I have not implemented the dataclass decorator in MicroPython. Instead I used the regular TypedDict, which is clean and clear enough for structuring data in my application.
